os/mm/mm_heap/mm_getheap.c

```diff
@@ -259,3 +259,48 @@ int mm_get_heapindex(void *mem)
 	/* The address does not lie in the user heaps */
 	return INVALID_HEAP_IDX;
 }
+
+static size_t mm_find_largest_freesize(struct mm_heap_s *heap)
+{
+	size_t largest_size = 0;
+	struct mm_freenode_s *fnode;
+	int i = 0;
+
+	for(i = 0; i < MM_NNODES; ++i) {
+		for (fnode = heap->mm_nodelist[i].flink; fnode && fnode->size; fnode = fnode->flink) {
+			if (largest_size < fnode->size) {
+				largest_size = fnode->size;
+			}
+		}
+	}
+	return largest_size;
+}
+
+size_t mm_get_largest_freesize(void *address)
+{
+	struct mm_heap_s *heap;
+	size_t largest_size = 0;
+	size_t ret;
+
+#ifndef __KERNEL__
+	heap = mm_get_heap(address);
+	if (!heap) {
+		mdbg("Fail to find the heap.\n");
+		return 0;
+	}
+	largest_size = mm_find_largest_freesize(heap);
+#else
+	int heap_idx;
+	heap = g_kmmheap;
+	struct mm_heap_s *target_heap;
+	for (heap_idx = HEAP_START_IDX; heap_idx <= HEAP_END_IDX; heap_idx++) {
+		target_heap = &heap[heap_idx];
+		ret = mm_find_largest_freesize(target_heap);
+		if (ret > largest_size) {
+			largest_size = ret;
+		}
```

Comment:
Could you let me know the meaning of this if conditional?

Reply:
Kernel can have multi-heap, so we should check whole heaps to find the largest.
So we can get the largest node size after comparing N-heaps.